lib/storage.hh

```diff
@@ -221,6 +221,175 @@ public:
 
 
 /*
+ * \class NibbleStorage
+ *
+ * \brief A A CountMin sketch implementation using 4bit counters.
+ *
+ * NibbleStorage is used to track counts of k-mers by Hashtable
+ * and derived classes.  It contains 'n_tables' different tables of
+ * 'tablesizes' entries. It allocates half a byte per table entry.
+ *
+ * Like other Storage classes, NibbleStorage manages setting the bits and
+ * tracking statistics, as well as save/load, and not much else.
+ *
+ */
+class NibbleStorage : public Storage
+{
+protected:
+    // table size is measured in number of entries in the table, not in bytes
+    std::vector<uint64_t> _tablesizes;
+    size_t _n_tables;
+    uint64_t _occupied_bins;
+    uint64_t _n_unique_kmers;
+    uint8_t _max_count{15};
+    Byte ** _counts;
+
+    // Compute index into the table, this retrieves the correct byte
+    // which you then need to select the correct nibble from
+    uint64_t _table_index(const HashIntoType k, const uint64_t tablesize) const
+    {
+        const uint64_t bins = tablesize / 2 + 1;
+        return (k / 2) % bins;
+    }
+    // Compute which half of the byte to use for this hash value
+    uint8_t _mask(const HashIntoType k) const
+    {
+        return k%2 ? 15 : 240;
```

Comment:
This may be premature optimization, but instead of k%2 what about just looking at the first bit?